### ISSUE TITLE
Breaks stacking and tabbed layouts on Sway 1.5.1

### ISSUE THREAD

**Issue by in0ni** [nwg-piotr/autotiling#20]
First of all, I have been looking for this solution and it works great! Thank you.

Though I've noticed that while this script is running, I am unable to switch layouts using my key bindings. Even after killing the script I can't get it to work. I have to comment `exec autotiling` and restart sway in order for these layouts to work.

I've tested both versions 1.1, and autotiling-git (aur). I am running sway 1.5.1. Any additional information I could provide, or help please let me know.

Thanks again :)

**Comment by nwg-piotr** [comment]
I hardly ever use stacking and tabbed layouts, so I've not noticed the issue so far. I gave it a try, and it seems to look a bit different here. Firstly: tabbed and stacked layouts (and also floating & fullscreen containers) are being [ignored](https://github.com/nwg-piotr/autotiling/blob/58de027b790d56515b202d4a26abe7fda9cd9dbc/autotiling/main.py#L46) by the script. No action is being taken. You may check it by running `autotiling --debug` from the terminal, instead of auto-starting it.

On my machine switching to the stacked layout happens to kill ALL key bindings (randomly, not every time!), **regardless of if autotiling is running or not**. Sometimes it happens, sometime it doesn't happen. IMO this may be a bug in sway, not in the autotiling script.

**Comment by in0ni** [comment]
@nwg-piotr -- thanks for the quick response. I've been testing this a little and I have some new findings that might provide some insight.

- Yes; stacked, tabbed, floating, and full screen are all ignored by the script
- I've done my testing using `autotiling --debug`, which has been very helpful

We're definitely experiencing different behavior. What version of sway/autotiling are you using?
- Fullscreen and floating keybindings work just fine, always
- I've tested a bunch of different keybindings, all of them work just fine (with the exception of tiling layout related bindings)
- I noticed that even after exiting the script, the keybindings for `stacking` and `tabbed` stop working ONLY on those windows that have had action taken by the script.
- I found how to get keybindings working after the script is terminated.

Here is a quick test that **I can replicate every time**:
- I have a screen (all tiling windows) with two containers, both with two child windows each (horizontally split)
- I can switch each container between tabbed, stacking without a problem (as usual)
- I can also toggle the split between horz/vert for each container without a problem (as usual)
- I start the script with `autotiling --debug` in another window (outside of both containers)
- I focus a single window -- and a single debug statement is printed
- I kill the script (with a single debug statement printed, so only one window was 'touched' by the script)
- The container where **no child window was touched** works as expected, can do all layout related actions
- The container where **a single child window was focused**, and 'touched' by the script is now 'broken':
  - Cannot switch between tabbed, stacking
  - Once any of the child windows (no matter if it's the focused one or not) is closed, and another is opened in the same container, it's all back to normal.

I have found a few other things, but I feel these are the most relevant.

**Comment by in0ni** [comment]
There's nothing like stepping away from a problem... the above tests got me thinking since I also noticed that I am unable to `toggle split` on the containers after a window within a container is 'touched' by the script...

After a quick test, **I found the "issue"**:  once a window has `splitv` or `splith` set on it, you are no longer able to toggle the split, switch to tabbed layout, or switch to stacking layout (default behavior, without the script running).

This kind of sucks, but it makes perfect sense with the current window workflow of sway: if you _consciously_ make a decision to set `splitv` or `splith`, then you clearly are sticking to tiled layout and are about to open a new window that you want positioned where you decided. This is why when new windows are opened they are not set to either -- you can tell because there is no default 'active' window border. Once a user selects which split they want, either the _bottom_ or _right_ window border is 'active' meaning it's ready to open a new window to the selected side.

**Possible Fix:** Do you think it's possible to catch the event for a new window being opened, and set `splitv` or `splith` right _before_ the window is opened? Instead of when it's focused? From my tests there is no problem by setting either option on a window right before opening a new one and then switching to tabbed or stacking layouts -- or toggling the split for that matter. All seems to work just fine.

Having `splitv` or `splith` set on any focused window is problematic and breaks the use of other layouts, as well as features like `toggle split`, which makes the script only usable for users who will never use any of those layout features :/

**Comment by nwg-piotr** [comment]
> Do you think it's possible to catch the event for a new window being opened, and set splitv or splith right before the window is opened? Instead of when it's focused?

The list of event we can subscribe to you may check [here](https://github.com/altdesktop/i3ipc-python/blob/master/i3ipc/events.py). Just choose one. ;)

**Comment by nwg-piotr** [comment]
My previous experience came from my development machine. Now I gave it a try on my ancient testing laptop. I was trying hard to reproduce the issue here, but I failed. With autotiling running, nothing stops me from switching layouts from key bindings, as well with `layout tabbed` / `layout stacking`, as with `layout toggle <layout>`.

**Comment by in0ni** [comment]
@nwg-piotr -- I'm a little confused about your results. Please note that I have described that the cause of the problem is due to how sway behaves (by design, not a bug), and can be easily be verified without your script. It's important to note this behavior as it directly affects your script, especially since **the approach taken by your script will always break switching layouts**. Perhaps my last post, which describes the problem, was too wordy :wink: 

Please take a look at [my post on reddit, where I try to get feedback](https://www.reddit.com/r/swaywm/comments/k2ibml/assistance_w_reproducing_an_issue_with_window/) from other sway users, and includes a quick easy way to reproduce _only using sway_. Especially take a look at the comments by one user.

I really would like the feature your script provides (like many other users), but sadly the way it works (setting the window to split on focus) _will always break layout functionality in sway_, or as stated by /u/sphutp2:

> "It seems like each window that is created is automatically pushed down into a nested container, which messes up switching layouts"

**Comment by nwg-piotr** [comment]
Well, I might have misunderstood something, but I still couldn't confirm the issue. See https://youtu.be/RFHy2TiI7sA

**Comment by in0ni** [comment]
Thanks for the video! Though it's quite hard to know what is going on without knowing exactly what keys you are pressing, the actual order of keys and flow is important here.

Are you able to replicate the steps mentioned in the reddit post? This is a _very simple and quick test_ -- this is _without_ autotiling running, just plain sway.

If you could replicate that test (without autotiling) maybe that would help clarify things? Cause so far for me, and the user on reddit, we're experiencing the same behavior which is expected.

**Comment by nwg-piotr** [comment]
W/o autotiling I can do all the steps you named in the reddit post. With autotiling on, two windows opened side by side won't become a single tabbed/stacking layout. I can only switch one of them at a time - and its children.

I'm on a triple-headed setup and never ever need tabbed or stacking layouts, but I see that it may be inconvenient. At the moment I have no idea on how it could be improved.

**Comment by in0ni** [comment]
I suggested of a possible fix on my third post, if it's possible. I saw the events, just not sure if you can capture it before it triggers.

**Comment by nwg-piotr** [comment]
This is super easy to check. Clone the repo, `cd your_clone_dir/autotiling/autotiling` and run `python main.py --debug`. Try changing the event in main.py line 90. None of what I tried seemed to give a desired result.

**Comment by nwg-piotr** [comment]
Recently introduced `--workspaces` argument by @riscie #22 is a solution to some degree. It allows to enable autotiling on certain workspaces only. I think nothing better can be done.
